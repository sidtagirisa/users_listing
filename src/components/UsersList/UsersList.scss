@import "../../styles/variables.scss";

.list {
  max-width: $max-width;
  margin: auto;
  margin-top: 2%;
  background: #fff;
  box-shadow: $box-shadow;
  width: 50%;
  padding: 35px 40px;
  z-index: 1;
}