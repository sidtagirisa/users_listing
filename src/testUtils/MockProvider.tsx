import React from 'react';
import StateProvider from '../context/StateProvider/index';
import initialState from '../context/initialState';

export default function MockProvider(props: {
  additionalStateRequired?: object;
  children: JSX.Element;
}): JSX.Element {
  const { additionalStateRequired, children } = props;

  return (
    <StateProvider
      initialState={{
        ...initialState,
        ...additionalStateRequired,
      }}
    >
      {children}
    </StateProvider>
  );
}

MockProvider.defaultProps = {
  value: {
    state: {},
    dispatch: jest.fn(),
  },
};
