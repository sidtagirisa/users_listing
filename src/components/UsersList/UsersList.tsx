import React, { useContext } from 'react';
import { User } from '../../context/initialState';
import { StateContext } from '../../context/StateProvider';
import './UsersList.scss';
import Info from '../Info';

export default function UsersList(props: { filterValue: string }): JSX.Element {
  // @ts-ignore
  const { state } = useContext(StateContext);
  const { users } = state;

  let filteredUsers = users.filter((o: any) =>
    Object.keys(o).some(
      (k) =>
        typeof o[k] === 'string' &&
        o[k].toLowerCase().includes(props.filterValue.toLowerCase())
    )
  );

  return (
    <React.Fragment>
      <div>
        {filteredUsers.map((user: User) => (
          <div className="list" key={user.id}>
            <Info heading="name" value={user.name} />
            <Info heading="username" value={user.username} />
            <Info heading="email" value={user.email} />
            <Info
              heading="address"
              value={`${user.address.suite}, ${user.address.street}, ${user.address.city}, ${user.address.zipcode}`}
            />
            <Info
              heading="phone"
              value={<a href={`tel:${user.phone}`}>{user.phone}</a>}
            />
            <Info
              heading="website"
              value={user.website}
              description={`${user.company.catchPhrase}`}
            />
          </div>
        ))}
      </div>
    </React.Fragment>
  );
}
