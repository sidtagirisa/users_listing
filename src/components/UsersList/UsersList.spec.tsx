import { render } from '@testing-library/react';
import React from 'react';
import UsersList from '.';
import { users } from '../../testUtils/fixtures';
import MockProvider from '../../testUtils/MockProvider';

describe('# UsersList', () => {
  it('should render list of users', async () => {
    const { container } = render(
      <MockProvider additionalStateRequired={{ loading: false, users }}>
        <UsersList filterValue="" />
      </MockProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not render anything if filtered value is not present', async () => {
    const { container } = render(
      <MockProvider additionalStateRequired={{ loading: false, users }}>
        <UsersList filterValue="some-key" />
      </MockProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
