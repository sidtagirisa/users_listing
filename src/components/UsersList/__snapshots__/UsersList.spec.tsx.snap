// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# UsersList should not render anything if filtered value is not present 1`] = `<div />`;

exports[`# UsersList should render list of users 1`] = `
<div>
  <div
    class="list"
  >
    <div
      class="information"
    >
      <h4
        class="heading"
      >
        name
      </h4>
      <p
        class="value"
      >
        Leanne Graham
      </p>
    </div>
    <div
      class="information"
    >
      <h4
        class="heading"
      >
        username
      </h4>
      <p
        class="value"
      >
        Bret
      </p>
    </div>
    <div
      class="information"
    >
      <h4
        class="heading"
      >
        email
      </h4>
      <p
        class="value"
      >
        Sincere@april.biz
      </p>
    </div>
    <div
      class="information"
    >
      <h4
        class="heading"
      >
        address
      </h4>
      <p
        class="value"
      >
        Apt. 556, Kulas Light, Gwenborough, 92998-3874
      </p>
    </div>
    <div
      class="information"
    >
      <h4
        class="heading"
      >
        phone
      </h4>
      <p
        class="value"
      >
        <a
          href="tel:1-770-736-8031 x56442"
        >
          1-770-736-8031 x56442
        </a>
      </p>
    </div>
    <div
      class="information"
    >
      <h4
        class="heading"
      >
        website
      </h4>
      <p
        class="value"
      >
        hildegard.org
      </p>
      <p
        class="description"
      >
        Multi-layered client-server neural-net
      </p>
    </div>
  </div>
</div>
`;
